examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+	{
+		LOG("Window gained keyboard focus");
+	}
+	void onLostKeyboardFocus_impl() override
+	{
+		LOG("Window lost keyboard focus");
+	}
+
+	void onMouseConnected_impl(core::smart_refctd_ptr<nbl::ui::IMouseEventChannel>&& mch) override
+	{
+		LOG("A mouse has been connected");
+	}
+	void onMouseDisconnected_impl(nbl::ui::IMouseEventChannel* mch) override
+	{
+		LOG("A mouse has been disconnected");
+	}
+	void onKeyboardConnected_impl(core::smart_refctd_ptr<nbl::ui::IKeyboardEventChannel>&& kbch) override
+	{
+		LOG("A keyboard has been connected");
+	}
+	void onKeyboardDisconnected_impl(nbl::ui::IKeyboardEventChannel* mch) override
+	{
+		LOG("A keyboard has been disconnected");
+	}
+};
 
 int main()
 {
 	constexpr uint32_t WIN_W = 800u;
 	constexpr uint32_t WIN_H = 600u;
 	constexpr uint32_t SC_IMG_COUNT = 3u;
-
-	auto win = CWindowT::create(WIN_W, WIN_H, system::IWindow::ECF_NONE);
-
+	
+	// Note(achal): This is unused, for now
 	video::SDebugCallback dbgcb;
 	dbgcb.callback = &debugCallback;
 	dbgcb.userData = nullptr;
-	auto gl = video::IAPIConnection::create(video::EAT_OPENGL, 0, "New API Test", &dbgcb);
-	auto surface = gl->createSurface(win.get());
+	
+	auto system = CommonAPI::createSystem();
+
+	auto winManager = core::make_smart_refctd_ptr<nbl::ui::CWindowManagerWin32>();
+
+	nbl::ui::IWindow::SCreationParams params;
+	params.callback = nullptr;
+	params.width = WIN_W;
+	params.height = WIN_H;
+	params.x = 0;
+	params.y = 0;
+	params.system = core::smart_refctd_ptr(system);
+	params.flags = nbl::ui::IWindow::ECF_NONE;
+	params.windowCaption = "Test Window";
+	params.callback = core::make_smart_refctd_ptr<DemoEventCallback>();
+	auto window = winManager->createWindow(std::move(params));
+
+	core::smart_refctd_ptr<video::IAPIConnection> vk = video::IAPIConnection::create(std::move(system), video::EAT_VULKAN, 0, "New API Test", &dbgcb);
+	core::smart_refctd_ptr<video::ISurface> surface = vk->createSurface(window.get());
+
+	auto gpus = vk->getPhysicalDevices();
+	assert(!gpus.empty());
+	
+	// Find a suitable gpu
+	// 
+	// Todo(achal): This process gets quite involved in Vulkan, do we want to delegate some of it to the engine.
+	// For example, the user just specifies if the application is a headless compute one and we could check the
+	// availability of the required extensions on the backend and report which physical device is suitable or
+	// some thing like that
+
+	// Todo(achal): Look at this:
+	// https://github.com/Devsh-Graphics-Programming/Nabla/blob/6bd5061abe0a2020142efda827269ea6c07f0f2f/examples_tests/common/CommonAPI.h
+	core::smart_refctd_ptr<video::CVulkanPhysicalDevice> gpu = nullptr;
+
+	uint32_t graphicsFamilyIndex(~0u);
+	uint32_t presentFamilyIndex(~0u);
+
+	// Todo(achal): Probably want to put these into some struct
+	nbl::video::ISurface::SFormat surfaceFormat;
+	nbl::video::ISurface::E_PRESENT_MODE presentMode;
+	nbl::video::ISurface::E_SURFACE_TRANSFORM_FLAGS preTransform;

```

Comment:
are there any limitations/capabilities in Vulkan on pretransform ?

For example, I know you can query which present modes are supported by a surface (which is really useful because OpenGL only supports 3 out of the 5 Vulkan ones #145 )

Reply:
`VkSwapchainCreateInfoKHR::preTransform` should be found by querying `VkSurfaceCapabilitiesKHR` and using its `currentTransform` member, failing to do so will produce a performance validation warning.

Reply:
ok, but is there no query which could say "You cannot use this transform!" ?

Reply:
yep there is, in `VkSurfaceCapabilitiesKHR`
> supportedTransforms is a bitmask of VkSurfaceTransformFlagBitsKHR indicating the presentation transforms supported for the surface on the specified device. At least one bit will be set.

Reply:
resolved?

Reply:
do you report them?

Reply:
not yet

Reply:
Yes, they're reported in surface caps now.
Resolved

Reply:
remember to implement on OpenGL backend.